Assets/Mirror/Runtime/NetworkClient.cs

```diff
@@ -223,59 +224,55 @@ internal void RegisterSystemHandlers(bool localClient)
+                RegisterHandler<ObjectDestroyMessage>(ClientScene.OnLocalClientObjectDestroy);
+                RegisterHandler<ObjectHideMessage>(ClientScene.OnLocalClientObjectHide);
+                RegisterHandler<OwnerMessage>((msg) => {});
+                RegisterHandler<NetworkPongMessage>((msg) => {});
+                RegisterHandler<SpawnPrefabMessage>(ClientScene.OnLocalClientSpawnPrefab);
+                RegisterHandler<SpawnSceneObjectMessage>(ClientScene.OnLocalClientSpawnSceneObject);
+                RegisterHandler<ObjectSpawnStartedMessage>((msg) => {});
+                RegisterHandler<ObjectSpawnFinishedMessage>((msg) => {});
+                RegisterHandler<UpdateVarsMessage>((msg) => {});
             }
             else
             {
-                RegisterHandler(MsgType.LocalClientAuthority, ClientScene.OnClientAuthority);
-                RegisterHandler(MsgType.ObjectDestroy, ClientScene.OnObjectDestroy);
-                RegisterHandler(MsgType.ObjectHide, ClientScene.OnObjectDestroy);
-                RegisterHandler(MsgType.Owner, ClientScene.OnOwnerMessage);
-                RegisterHandler(MsgType.Pong, NetworkTime.OnClientPong);
-                RegisterHandler(MsgType.SpawnPrefab, ClientScene.OnSpawnPrefab);
-                RegisterHandler(MsgType.SpawnSceneObject, ClientScene.OnSpawnSceneObject);
-                RegisterHandler(MsgType.SpawnStarted, ClientScene.OnObjectSpawnStarted);
-                RegisterHandler(MsgType.SpawnFinished, ClientScene.OnObjectSpawnFinished);
-                RegisterHandler(MsgType.UpdateVars, ClientScene.OnUpdateVarsMessage);
+                RegisterHandler<ObjectDestroyMessage>(ClientScene.OnObjectDestroy);
+                RegisterHandler<ObjectHideMessage>(ClientScene.OnObjectHide);
+                RegisterHandler<OwnerMessage>(ClientScene.OnOwnerMessage);
+                RegisterHandler<NetworkPongMessage>(NetworkTime.OnClientPong);
+                RegisterHandler<SpawnPrefabMessage>(ClientScene.OnSpawnPrefab);
+                RegisterHandler<SpawnSceneObjectMessage>(ClientScene.OnSpawnSceneObject);
+                RegisterHandler<ObjectSpawnStartedMessage>(ClientScene.OnObjectSpawnStarted);
+                RegisterHandler<ObjectSpawnFinishedMessage>(ClientScene.OnObjectSpawnFinished);
+                RegisterHandler<UpdateVarsMessage>(ClientScene.OnUpdateVarsMessage);
             }
-
-            RegisterHandler(MsgType.Rpc, ClientScene.OnRPCMessage);
-            RegisterHandler(MsgType.SyncEvent, ClientScene.OnSyncEventMessage);
+            RegisterHandler<ClientAuthorityMessage>(ClientScene.OnClientAuthority);
+            RegisterHandler<RpcMessage>(ClientScene.OnRPCMessage);
+            RegisterHandler<SyncEventMessage>(ClientScene.OnSyncEventMessage);
         }
 
-        public void RegisterHandler(short msgType, NetworkMessageDelegate handler)
+        public void RegisterHandler<T>(Action<T> handler) where T : MessageBase, new()
```

Comment:
why not use Action<NetworkConnection, T> for clients too?
for example, in uMMORPG I have a OnClientError(ErrorMessage msg) function which should disconnect the client's own connection after an error. previously I could call netMsg.conn.Disconnect(). Now I have to call NetworkClient.singleton.connection.Disconnect() which is more annoying.
Passing connection as parameter would be easier to use imho.
probably easier to upgrade for others too.

Reply:
How about a static connection property on NetworkClient to get closer to the eventuality of NetworkClient becoming a static class?

Reply:
fixed